### PR TITLE
Org chooser with avatars

### PR DESCRIPTION
**Pull request by ericnewcomer**
* Uses new _popup_  and _avatar_ options on temba-menu
 * Avatars are created using a color based on a hash of the name and shows two characters (either of two words or two first of a single word). Eventually perhaps we'll allow accounts to have an optional image to represent them.
 * Workspace management page with create / delete
 * Some small tweaks to the dashboard
 * Moves sign out to be in popup menu

Screenshot 2023-01-26 at 11 40 53 AM
Screenshot 2023-01-26 at 11 42 39 AM
Screenshot 2023-01-26 at 11 41 50 AM
Screenshot 2023-01-26 at 11 46 31 AM

### CODE REVIEW

**Comment by ericnewcomer** [comment]
@rowanseymour was hoping to ship this tomorrow am

File: temba/dashboard/tests.py [hunk 1]
```diff
@@ -48,18 +48,15 @@ def test_message_history(self):
         self.create_activity()
         response = self.client.get(url).json()
 
-        # in, out, and total
-        self.assertEqual(3, len(response))
+        # in, out
+        self.assertEqual(2, len(response))
```
Comment (ericnewcomer) [inline temba/dashboard/tests.py:52]: Using the new highcharts now which doesn't necessitate that we bring over a _total_ series, so it's been removed

File: temba/orgs/views.py [hunk 2]
```diff
@@ -1119,14 +1120,17 @@ def create_menu_item(
...
         endpoint=None,
         href=None,
         count=None,
         perm=None,
         items=[],
         inline=False,
         bottom=False,
+        popup=False,
         verbose_name=None,
+        event=False,
```
Comment (ericnewcomer) [inline temba/orgs/views.py:1133]: `avatar` is a url to an image or a test string to build an avatar from
`popup` means the menu will overlay using an internal `temba-dropdown`
`event` means is `temba-menu` won't deal with it for you and expect event handling in the page to handle it

File: temba/orgs/views.py [hunk 3]
```diff
@@ -2400,27 +2455,10 @@ def derive_title(self):
...
-                    )
-
-                    if self.has_org_perm("orgs.org_delete"):
-                        menu.add_modax(
-                            _("Delete"),
-                            "delete-workspace",
-                            reverse("orgs.org_delete", args=[self.object.id]),
-                            title=_("Delete Workspace"),
-                            disabled=True,
-                        )
```
Comment (ericnewcomer) [inline temba/orgs/views.py:2421]: These are now managed from the sub orgs list page. We don't rely on doing this from the user manage page any longer which was always a bit weird.

File: temba/orgs/views.py [hunk 4]
```diff
@@ -2536,12 +2580,9 @@ def get_contacts(self, obj):
             return obj.get_contact_count()
 
         def get_name(self, obj):
-            org_type = "child"
-            if not obj.parent:
-                org_type = "parent"
```
Comment (ericnewcomer) [inline temba/orgs/views.py:2541]: The sub orgs page only has... well, sub orgs. No longer for a special cased parent (which was only there for transferring credits back in the day)

File: temba/utils/templatetags/temba.py [hunk 5]
```diff
@@ -226,6 +226,11 @@ def datetime(date):
     return mark_safe(f"<temba-date value='{date.isoformat()}' display='datetime'></temba-date>")
 
 
+@register.filter
+def day(date):
+    return mark_safe(f"<temba-date value='{date.isoformat()}' display='date'></temba-date>")
```
Comment (ericnewcomer) [inline temba/utils/templatetags/temba.py:231]: which we could use date here, but that is overloaded

File: templates/spa_frame.haml [hunk 6]
```diff
@@ -472,16 +474,6 @@
...
-
-        menu.addEventListener("temba-button-clicked", function(event) {
-          var modaxOptions = { 
-            disabled: false, 
-            onSubmit: event.detail.on_submit
-          }
-
-          showModax(event.detail.title, event.detail.href, modaxOptions);
-        });
-
```
Comment (ericnewcomer) [inline templates/spa_frame.haml:484]: These are now handled as an `event` type instead

File: temba/orgs/tests.py [hunk 7]
```diff
@@ -2850,11 +2850,17 @@ def test_spa(self):
...
+    def assertMenu(self, url, count, checks=[]):
         response = self.assertListFetch(url, allow_viewers=True, allow_editors=True, allow_agents=True)
         menu = response.json()["results"]
         self.assertEqual(count, len(menu))
 
+        # check the content if we have them
+        if checks:
+            content = json.dumps(menu, indent=2)
+            for check in checks:
+                self.assertTrue(check in content, f"Couldn't find {check} in {content}")
```
Comment (rowanseymour) [inline temba/orgs/tests.py:2862]: `self.assertIn(check, content)`

File: temba/orgs/tests.py [hunk 8]
```diff
@@ -2850,11 +2850,17 @@ def test_spa(self):
         response = self.client.get(deep_link)
         self.assertEqual(200, response.status_code)
 
-    def assertMenu(self, url, count):
+    def assertMenu(self, url, count, checks=[]):
```
Comment (rowanseymour) [inline temba/orgs/tests.py:2853]: in `CRUDLTestMixin` methods `checks` means specifically the check objects.. so maybe `contains`

File: temba/orgs/tests.py [hunk 9]
```diff
@@ -2850,11 +2850,17 @@ def test_spa(self):
...
         response = self.assertListFetch(url, allow_viewers=True, allow_editors=True, allow_agents=True)
         menu = response.json()["results"]
         self.assertEqual(count, len(menu))
 
+        # check the content if we have them
+        if checks:
+            content = json.dumps(menu, indent=2)
```
Comment (rowanseymour) [inline temba/orgs/tests.py:2860]: if we're checking menu titles then better to make this assert more specific like test for membership in `[m["title"] for m in menu]`

File: temba/utils/templatetags/temba.py (hunk 5 above)
Comment (rowanseymour) [reply temba/utils/templatetags/temba.py:231]: you mean that django has a filter called `date`? Can we just override it since we don't want its behavior?

**Review by rowanseymour — CHANGES_REQUESTED** [review]

File: temba/orgs/tests.py (hunk 9 above)
Comment (ericnewcomer) [reply temba/orgs/tests.py:2860]: Also looking for parts of urls.. You want me to pull out and check both?

File: temba/utils/templatetags/temba.py (hunk 5 above)
Comment (ericnewcomer) [reply temba/utils/templatetags/temba.py:231]: Tried, got angry.

File: temba/orgs/tests.py (hunk 9 above)
Comment (rowanseymour) [reply temba/orgs/tests.py:2860]: Not seeing any calls checking parts of URLs...

Comment (ericnewcomer) [reply temba/orgs/tests.py:2860]: hmmm.. guess i changed that test 🤷

Comment (ericnewcomer) [reply temba/orgs/tests.py:2860]: ...oh, and this is look in a child list of items, and the depth can be arbitrary. can unwind that but feels unnecessarily complicated.

Comment (rowanseymour) [reply temba/orgs/tests.py:2860]: ok n/m - tho see comment above and below

File: temba/orgs/tests.py [hunk 10]
```diff
@@ -2850,11 +2850,25 @@ def test_spa(self):
...
+                steps = name.split("/")
+                while steps:
+                    step = steps.pop(0)
+                    menu_names = [m["name"] for m in menu if "name" in m]
+                    try:
+                        idx = menu_names.index(step)
+                        if "items" in menu[idx]:
+                            menu = menu[idx]["items"]
+                    except ValueError:
+                        self.fail(f"Couldn't find {step} in {menu_names}")
```
Comment (ericnewcomer) [inline temba/orgs/tests.py:2870]: @rowanseymour switched it to force a named path to be sent, ie `self.assertMenu(..., ["Workspace/Child Workspace"])`

Comment (rowanseymour) [reply temba/orgs/tests.py:2870]: neato

**Review by rowanseymour — APPROVED** [review]

**Pull request merged**